### PR TITLE
Default `kubernetes.version` if omitted

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area usability
/kind enhancement

**What this PR does / why we need it**:
Gardener already allows to partially specify Kubernetes versions, e.g. `1.27` and defaults it the highest supported patch version found in the linked CloudProfile `1.27 -> 1.27.3`.
This functionality is enhanced by the following changes, so that users can completely omit the Kubernetes version in their shoot specs. Gardener will automatically default them to the latest, supported and not expired version, e.g. `"" -> 1.27.3`, `1 -> 1.27.3`.

**Release note**:

```feature user
Gardener now allows to omit or to only partially define Kubernetes versions in `Shoot`s. The version will automatically be defaulted to the latest minor and/or patch version found in the linked `CloudProfile`.
```

### CODE REVIEW

File: plugin/pkg/shoot/validator/admission.go [hunk 1]
```diff
@@ -687,15 +687,16 @@ func (c *validationContext) validateKubernetes(a admission.Attributes) field.Err
...
-		c.shoot.Spec.Kubernetes.Version = versionDefault.String()
+	defaultVersion, errList := defaultKubernetesVersion(c.cloudProfile.Spec.Kubernetes.Versions, c.shoot.Spec.Kubernetes.Version, path.Child("version"))
+	if len(errList) > 0 {
+		allErrs = append(allErrs, errList...)
+	}
+
+	if defaultVersion != nil {
+		c.shoot.Spec.Kubernetes.Version = *defaultVersion
+	} else {
+		// We assume that the 'defaultVersion' is already calculated correctly, so only run validation if the verion was not defaulted.
```
Comment (rfranzke) [inline plugin/pkg/shoot/validator/admission.go:771]:
```suggestion
		// We assume that the 'defaultVersion' is already calculated correctly, so only run validation if the version was not defaulted.
```

File: pkg/apis/core/types_shoot.go [hunk 2]
```diff
@@ -431,6 +431,8 @@ type Kubernetes struct {
 	// Kubelet contains configuration settings for the kubelet.
 	Kubelet *KubeletConfig
 	// Version is the semantic Kubernetes version to use for the Shoot cluster.
+	// Defaults to the highest supported minor and patch version given in the referenced cloud profile.
+	// The version can be omitted completely or partially specified, e.g. `<major>.<minor>`.
 	Version string
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go:440]: pointer?

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 3]
```diff
@@ -531,7 +531,10 @@ type Kubernetes struct {
 	// +optional
 	Kubelet *KubeletConfig `json:"kubelet,omitempty" protobuf:"bytes,7,opt,name=kubelet"`
 	// Version is the semantic Kubernetes version to use for the Shoot cluster.
-	Version string `json:"version" protobuf:"bytes,8,opt,name=version"`
+	// Defaults to the highest supported minor and patch version given in the referenced cloud profile.
+	// The version can be omitted completely or partially specified, e.g. `<major>.<minor>`.
+	// +optional
+	Version string `json:"version,omitempty" protobuf:"bytes,8,opt,name=version"`
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go:541]: pointer?

**Review by rfranzke — CHANGES_REQUESTED** [review]
> Defaulting logic that relies on the Kubernetes version must run after the Kubernetes version has been set. This code can't remain in the API defaulting area since this part if processed before the admission plugins.

How do we make sure we don't forget about this?

**Comment by timuthy** [comment]
> How do we make sure we don't forget about this?

Existing e2e tests will fail w/o this change. Do you have anything particular in mind?

**Comment by rfranzke** [comment]
> > How do we make sure we don't forget about this?
> 
> Existing e2e tests will fail w/o this change. Do you have anything particular in mind?

Maybe a note in https://github.com/gardener/gardener/blob/master/docs/development/new-kubernetes-version.md as well?

File: plugin/pkg/shoot/validator/admission.go (hunk 1 above)
Comment (timuthy) [reply plugin/pkg/shoot/validator/admission.go:771]: Please see eb3963710042358feabb98121d3c2833928bb9d2 😓 
I also tested this change WRT backwards compatibility and Protobuf: An "older" Gardenlet version was still able to connect to a "new" Apiserver ✅

File: pkg/apis/core/types_shoot.go (hunk 2 above)
Comment (timuthy) [reply pkg/apis/core/types_shoot.go:440]: Please see eb3963710042358feabb98121d3c2833928bb9d2 😓 
I also tested this change WRT backwards compatibility and Protobuf: An "older" Gardenlet version was still able to connect to a "new" Apiserver ✅

File: example/provider-local/shoot.yaml [hunk 4]
```diff
@@ -26,7 +26,6 @@ spec:
       maxSurge: 1
       maxUnavailable: 0
   kubernetes:
-    version: 1.27.1
```
Comment (shafeeqes) [inline example/provider-local/shoot.yaml:29]: Please adapt `example/provider-local/shoot-workerless.yaml` as well.

Comment (timuthy) [reply example/provider-local/shoot.yaml:29]: Hm, the file was already adapted 🤔

File: pkg/apis/core/types_shoot.go (hunk 2 above)
Comment (timuthy) [reply pkg/apis/core/types_shoot.go:440]: I dropped the commit again after our discussion and added a note why we stick to `string` 🙂

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**